### ISSUE TITLE
Jit compiler for is_training arguments 

### ISSUE THREAD

**Issue by VIGNESHinZONE** [deepmind/dm-haiku#187]
Hi, 

I'm trying to build a Haiku model which takes the argument `is_training` for different behavior during training and test time.

Here is just a simple example with dropout but in reality, I want to be using [Haiku's batch_norm](https://github.com/deepmind/dm-haiku/blob/master/haiku/_src/batch_norm.py#L122#L200) which takes the `is_training` argument

```
class Net(hk.Module):
  def __init__(self, output_size: int = 1):
    super().__init__()
    self._network1 = hk.Sequential([hk.Linear(200), jax.nn.relu])
    self.output = hk.Linear(output_size)

  def __call__(self, x, is_training):
    x = self._network1(x)
    if is_training:
       x = hk.dropout(hk.next_rng_key(), 0.1, x)
    output = self.output(x)
    return output

def f(x, is_training = True):
  net = Net(1)
  return net(x)

```
where `f` is the forward_fn (which i would later alias for `model.apply` ) .
According to Jax's [The Sharp Bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#control-flow) , we use `static_argnums` in case we want to execute the code with JIT compiler and also dont want to keep track confitional arguments (in this case `is_training`)

And hence Im currently using this update function -
```
# not tracing is_training argument
@functools.partial(jax.jit,  static_argnums=(4,))
def model_loss(params, rng,  target, x, is_training ):
  predict = forward_fn(params, rng, x, is_training)
  return loss(predict, target)

# not tracing is_training argument
@functools.partial(jax.jit,  static_argnums=(5,))
def update(opt_state, params, rng, target, x, is_training):
  batch_loss, grads = jax.value_and_grad(model_loss)(params, rng, target, x, is_training)
  updates, opt_state = optimizer.update(grads, opt_state)
  new_params = optax.apply_updates(params, updates)
  return new_params, opt_state

```

As you can see above, in order to JIT compile the above model, I'm passing `static_argnums` to run the model and if I dont it throws a `jax expr trace error`. I understand this error occurs only when we try to speed up the process with JIT

So here is my question - 

1) Is there a way in which I can run the above model with JIT compiled without passing `static_argnums`?
2) While using MLP network, im able to use it with JIT despite `dropout` taking either of the two arguments `[float, None]`. Im confused as to why this line is not throwing an trace error? 
https://github.com/deepmind/dm-haiku/blob/4ae60fd4fd2da3b2f8f9ad3ec6dfd893745b483b/haiku/_src/nets/mlp.py#L114 

**PS**: We are building a general Haiku/Jax framework inside [Deepchem](https://github.com/deepchem/deepchem) for applications of PINNs in biological & physical processes. So it would be really helpful if you suggest such that we could bypass the `static_argnums` arguments.

Thanks in advance,

**Comment by tomhennigan** [comment]
Hi @VIGNESHinZONE , it is typical in an update function to always apply dropout, so I would suspect you can simply remove this argument from your update function and pass it directly to your networks apply function. We have a similar "is_training" setup in our resnet example (for updating batchnorm stats), you may find it useful to review that https://github.com/deepmind/dm-haiku/blob/main/examples/imagenet/train.py#L148.

In general, if you have an is_training (or similar) argument, and you want to run your function with the value set to True and False, you will need to compile twice. You can achieve this using static_argums, or you can create and @jit two different functions and inside those functions pass `is_training=True` explicitly. This second option is what we do in the imagenet example I linked earlier (`train_step` is compiled and calls with `is_training=True`, `eval_batch` is compiled and calls with `is_training=False`).

Re 2, the key thing is that this comparison is using `is not` and not `!=`. This is not an operation JAX is involved in computing (checking the identity of two objects is not something Python lets you overload) so we don't get tracer errors. Another way of thinking about it is that this comparison does not rely on the numerical value of the array, just static information (e.g. is this None or something else) so it is safe to run this code under JIT.
